File: backend/auth-service/src/services/authService.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const userRepository = require('../repositories/userRepository');
const { publicUser } = require('../dto/authDto');

const jwtSecret = process.env.JWT_SECRET || 'change-this-secret';

function signToken(user) {
  return jwt.sign(
    { sub: user.id, email: user.email, role: user.role, name: user.name },
    jwtSecret,
    { expiresIn: '8h' }
  );
}

async function register({ name, email, password, role }) {
  if (!name || !email || !password) {
    const error = new Error('Nombre, email y password son obligatorios.');
    error.status = 400;
    throw error;
  }

  const existing = await userRepository.findByEmail(email);
  if (existing) {
    const error = new Error('El email ya esta registrado.');
    error.status = 409;
    throw error;
  }

  const passwordHash = await bcrypt.hash(password, 10);
  const user = await userRepository.createUser({
    name,
    email,
    passwordHash,
    role: role === 'ADMIN' ? 'ADMIN' : 'USER'
  });

  return { user: publicUser(user), token: signToken(user) };
}

async function login({ email, password }) {
  const user = await userRepository.findByEmail(email || '');
  if (!user || !(await bcrypt.compare(password || '', user.password_hash))) {
    const error = new Error('Credenciales invalidas.');
    error.status = 401;
    throw error;
  }

  return { user: publicUser(user), token: signToken(user) };
}

async function getProfile(id) {
  const user = await userRepository.findById(id);
  if (!user) {
    const error = new Error('Usuario no encontrado.');
    error.status = 404;
    throw error;
  }
  return publicUser(user);
}

module.exports = { register, login, getProfile };

File: backend/auth-service/src/main/java/com/ringbox/auth/web/AuthService.java
package com.ringbox.auth.web;

import com.ringbox.auth.jwt.JwtService;
import com.ringbox.auth.user.User;
import com.ringbox.auth.user.UserRepository;
import com.ringbox.auth.web.dto.AuthResponse;
import com.ringbox.auth.web.dto.LoginRequest;
import com.ringbox.auth.web.dto.RegisterRequest;
import com.ringbox.auth.web.dto.UserDto;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class AuthService {

  private final UserRepository users;
  private final PasswordEncoder passwordEncoder;
  private final JwtService jwtService;

  public AuthService(UserRepository users, PasswordEncoder passwordEncoder, JwtService jwtService) {
    this.users = users;
    this.passwordEncoder = passwordEncoder;
    this.jwtService = jwtService;
  }

  public AuthResponse register(RegisterRequest req) {
    if (req.name() == null || req.name().isBlank()
        || req.email() == null || req.email().isBlank()
        || req.password() == null || req.password().isBlank()) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Nombre, email y password son obligatorios.");
    }
    String email = req.email().trim().toLowerCase();
    if (users.findByEmailIgnoreCase(email).isPresent()) {
      throw new ResponseStatusException(HttpStatus.CONFLICT, "El email ya esta registrado.");
    }
    User user = new User();
    user.setName(req.name().trim());
    user.setEmail(email);
    user.setPasswordHash(passwordEncoder.encode(req.password()));
    user.setRole("ADMIN".equalsIgnoreCase(req.role()) ? "ADMIN" : "USER");
    user = users.save(user);
    return new AuthResponse(toDto(user), jwtService.createToken(user));
  }

  public AuthResponse login(LoginRequest req) {
    String email = req.email() == null ? "" : req.email().trim().toLowerCase();
    User user =
        users
            .findByEmailIgnoreCase(email)
            .filter(u -> passwordEncoder.matches(req.password() == null ? "" : req.password(), u.getPasswordHash()))
            .orElseThrow(
                () -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Credenciales invalidas."));
    return new AuthResponse(toDto(user), jwtService.createToken(user));
  }

  public UserDto getProfile(long userId) {
    User user =
        users
            .findById(userId)
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Usuario no encontrado."));
    return toDto(user);
  }

  private static UserDto toDto(User u) {
    return new UserDto(u.getId(), u.getName(), u.getEmail(), u.getRole());
  }
}
